### ISSUE TITLE
Support :scope CSS pseudo selector

### ISSUE THREAD

**Issue by atifaziz** [AngleSharp/AngleSharp#347]
Seems like there's no support currently built-in for the [`:scope` pseudo selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:scope). It's not listed in [`PseudoClassNames`](https://github.com/AngleSharp/AngleSharp/blob/19f5974adb85724edd5bdb7c5a26a581a9f478f6/src/AngleSharp/Css/PseudoClassNames.cs) and using it with `QuerySelectorAll`/`QuerySelector` yields neither an error nor the expected result. At the very least, an error should be thrown otherwise one is left wondering. At the very most, consider adding support. :) Since MDN lists `:scope` as _experimental_, it won't be shocking if this is put on the back burner but bear in mind that it is especially useful when it comes to selecting direct children of an element.

**Comment by FlorianRappl** [comment]
Hm sure that no error is thrown? Sounds strange, I'll look into it.

Otherwise, feel free to PR an implementation!

**Comment by atifaziz** [comment]
> Hm sure that no error is thrown? Sounds strange, I'll look into it.

Here's a quick test I tried using the latest version (0.9.6):

``` c#
var html = new WebClient().DownloadString("http://www.example.com/");
var doc = new HtmlParser().Parse(html);
var paras = doc.QuerySelector("body")
               .QuerySelectorAll(":scope > p");
```

No error thrown and `paras` is an empty sequence at the end.

**Comment by FlorianRappl** [comment]
One tip though. If you really want to download websites reliably you should use AngleSharp for it. Otherwise there is no guarantee on encoding correctness. Also you get better performance as AngleSharp will parse the source _while_ its being downloaded and not _after_ its being finished. This is especially true with chunked responses.

**Comment by atifaziz** [comment]
@FlorianRappl [I am aware of the encoding issues with `WebClient.DownloadString`](https://bitbucket.org/raboof/mannex/src/9a009bd95b556a769a6ca719afb8e6122550e7b7/src/Net/WebClient.cs?at=default&fileviewer=file-view-default#WebClient.cs-72:95) however it's good to know that AngleSharp can parse the document while being downloaded. Frankly, I'm not completely familiar with the entire API so wanted to keep the code dead simple to illustrate the CSS selector I tried and that lead to this issue.

**Comment by FlorianRappl** [comment]
Sure, no problem! This is the code I would have used (yes, I would normally write it in 4-lines, but I tried to keep it as clean and concise as your example).

``` cs
var context = BrowsingContext.New(Configuration.Default.WithDefaultRequester());
var document = await context.OpenAsync("http://www.example.com/");
var paras = doc.QuerySelector("body").QuerySelectorAll(":scope > p");
```

Actually, for having a proper unit test I will replace the external request, i.e., we have

``` cs
var document = await context.OpenAsync(res => res.Content("... content"));
```

**Comment by atifaziz** [comment]
@FlorianRappl Cheers for sharing the download code. :beers: If I end up using AngleSharp, it will be just for the parser, DOM and reading nodes.

**Comment by FlorianRappl** [comment]
This affected pseudo elements as well. This was actually quite a big bug in my opinion. Thanks for reporting this! The bug is fixed, I'll see if `:scope` can make it in v0.9.7 or if that will be handled in a separate issue for v0.9.8.

**Comment by FlorianRappl** [comment]
Done for the moment, the full impl. comes in v0.9.8 with #352.
